HARK/ConsumptionSaving/ConsIndShockFastModel.py

```diff
@@ -0,0 +1,427 @@
+    PerfForesightConsumerType,
+    ValueFunc,
+    MargValueFunc,
+    MargMargValueFunc,
+)
+from HARK.interpolation import LinearInterp
+from HARK.utilities import (
+    CRRAutility,
+    CRRAutilityP,
+    CRRAutilityPP,
+    CRRAutilityP_inv,
+    CRRAutility_invP,
+    CRRAutility_inv,
+    CRRAutilityP_invP,
+)
+
+__all__ = [
+    "PerfForesightSolution",
+    "ConsPerfForesightFastSolver",
+    "PerfForesightConsumerType",
+]
+
+utility = CRRAutility
+utilityP = CRRAutilityP
+utilityPP = CRRAutilityPP
+utilityP_inv = CRRAutilityP_inv
+utility_invP = CRRAutility_invP
+utility_inv = CRRAutility_inv
+utilityP_invP = CRRAutilityP_invP
+
+
+# =====================================================================
+# === Classes that help solve consumption-saving models ===
+# =====================================================================
+
+
+class PerfForesightSolution(HARKobject):
+    """
+    A class representing the solution of a single period of a consumption-saving
+    perfect foresight problem.
+
+    Here and elsewhere in the code, Nrm indicates that variables are normalized
+    by permanent income.
+    """
+
+    distance_criteria = ["cNrm", "mNrm"]
+
+    def __init__(
+        self,
+        mNrm=np.array([0.0, 1.0]),
+        cNrm=np.array([0.0, 1.0]),
+        vFuncNvrsSlope=0.0,
+        mNrmMin=0.0,
+        hNrm=0.0,
+        MPCmin=1.0,
+        MPCmax=1.0,
+    ):
+        """
+        The constructor for a new ConsumerSolution object.
+
+        Parameters
+        ----------
+        mNrm: np.array
+            (Normalized) corresponding market resource points for interpolation.
+        cNrm : np.array
+            (Normalized) consumption points for interpolation.
+        vFuncNvrsSlope: float
+            Constant slope of inverse value vFuncNvrs
+        mNrmMin : float
+            The minimum allowable market resources for this period; the consump-
+            tion function (etc) are undefined for m < mNrmMin.
+        hNrm : float
+            Human wealth after receiving income this period: PDV of all future
+            income, ignoring mortality.
+        MPCmin : float
+            Infimum of the marginal propensity to consume this period.
+            MPC --> MPCmin as m --> infinity.
+        MPCmax : float
+            Supremum of the marginal propensity to consume this period.
+            MPC --> MPCmax as m --> mNrmMin.
+
+        Returns
+        -------
+        None
+        """
+        self.mNrm = mNrm
+        self.cNrm = cNrm
+        self.vFuncNvrsSlope = vFuncNvrsSlope
+        self.mNrmMin = mNrmMin
+        self.hNrm = hNrm
+        self.MPCmin = MPCmin
+        self.MPCmax = MPCmax
+
+
+# =====================================================================
+# === Classes and functions that solve consumption-saving models ===
+# =====================================================================
+
+
+@njit(cache=True)
+def searchSSfunc(m, Rfree, PermGroFac, mNrm, cNrm, ExIncNext):
```

Comment:
For code-style and documentation needs, can you make these independent helper functions start with `_`.

Example:
``` python
def _searchSSfunc(.........):

def _addSSmNrmNumba(........):

def _solveConsPerfForesightNumba(........):

```
By doing this we can make sure we don't pollute the local namespace, documentation on ReadTheDocs and functions that users won't be accessing directly should start with `_`.